### PR TITLE
Simplification of Linq Any method

### LINKED ISSUE #12193: Simplification of Linq Any method for performance improvement
### NuGet Product(s) Affected

NuGet.exe

### Current Behavior

The program calls .Where() with a predicate and then uses Any to filter it out

### Desired Behavior

It should call Any directly, which would improve performance.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by SiliconA-Z**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12193

Regression? No

## Description

The current code is inefficient, where we use "Where().Any()" when "Any()" applies better.

## PR Checklist

- [X] PR has a meaningful title
- [X] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Common/SelfUpdater.cs [hunk 1]
```diff
@@ -121,7 +121,7 @@ internal async Task UpdateSelfFromVersionAsync(string exePath, bool prerelease,
 
                         using (Stream fromStream = await downloadResourceResult.PackageReader.GetStreamAsync(nugetExeInPackageFilePath, cancellationToken), toStream = File.Create(exePath))
                         {
-                            fromStream.CopyTo(toStream);
+                            await fromStream.CopyToAsync(toStream);
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/Common/SelfUpdater.cs]: The project `NuGet.CommandLine` is what gets compiled into `nuget.exe`, which only runs on the .NET Framework, not .NET (Core). From all my testing of async file access on .NET Framework, async file IO is "considerably" (relatively speaking, from benchmarks) slower than sync file IO. Granted, this code path will only copy a single file, so it won't be measurably different to humans. But the point is that async file IO is not universally better.

File: src/NuGet.Core/NuGet.Packaging/Signing/Utility/SigningUtility.cs [hunk 2]
```diff
@@ -272,7 +272,7 @@ public static async Task SignAsync(SigningOptions options, SignPackageRequest si
             {
                 if (unsignedPackageStream != null && !ReferenceEquals(unsignedPackageStream, options.InputPackageStream))
                 {
-                    unsignedPackageStream.Dispose();
+                    await unsignedPackageStream.DisposeAsync();
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Signing/Utility/SigningUtility.cs]: According to [the docs for `Stream.DisposeAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream.disposeasync?view=net-6.0), this API was added in .NET Core 3.0, which means it won't compile on .NET Framework.

**Review by zivkan — COMMENTED** [review]
Thank you for taking the time to make a contribution!

Unfortunately, the idealism in switching some APIs from sync to async are not in scope for the title of this PR and not mentioned in the PR description, but also I'm confident will cause compile time errors on .NET Framework. I suggest removing all the async changes, keeping the PR scoped to what the title says, and that will make it easier for us to accept the PR.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/RestoreCommandHandler.cs [hunk 3]
```diff
@@ -47,7 +47,7 @@ private async Task RunSolutionRestoreAsync()
                 SB.IBrokeredServiceContainer serviceContainer = await _asyncServiceProvider.GetServiceAsync<SB.SVsBrokeredServiceContainer, SB.IBrokeredServiceContainer>();
                 IServiceBroker serviceBroker = serviceContainer.GetFullAccessServiceBroker();
 
-                INuGetSolutionService nugetSolutionService = await serviceBroker.GetProxyAsync<INuGetSolutionService>(NuGetServices.SolutionService);
+                INuGetSolutionService nugetSolutionService = await serviceBroker.GetProxyAsync<INuGetSolutionService>(NuGetServices.SolutionService, cancellationToken: cancellationTokenSource.Token);
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/RestoreCommandHandler.cs]: Maybe better to make separate PR for this change, not related to `.Any()` change.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/IPackTaskRequest.cs [hunk 4]
```diff
@@ -13,7 +13,7 @@ namespace NuGet.Build.Tasks.Pack
     /// <typeparam name="TItem">
     /// The item type. This will either be <see cref="ITaskItem"/> or a <see cref="IMSBuildItem"/>.
     /// </typeparam>
-    public interface IPackTaskRequest<TItem>
+    public interface IPackTaskRequest<out TItem>
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/IPackTaskRequest.cs]: Is this change related to `.Any()` ?

**Comment by SiliconA-Z** [comment]
Fixed!

**Comment by erdembayar** [comment]
> Fixed!

I queued your PR for CI build, let's wait.

**Comment by erdembayar** [comment]
Thanks for your contribution.
It makes code more readable, but technically doesn't improve performance. I believe it still create same Linq tree for evaluation regardless of this change.

**Comment by nkolev92** [comment]
> Thanks for your contribution. It makes code more readable, but technically doesn't improve performance. I believe it still create same Linq tree for evaluation regardless of this change.

I ran some quick benchmarks: 

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Benchmarks
{
    [SimpleJob(RuntimeMoniker.Net60)]
    [MemoryDiagnoser]
    public class AnyBenchmark
    {
        Random random = new Random();
        [GlobalSetup]
        public void GlobalSetup()
        {
        }

        [Benchmark]
        public void Any()
        {
            var list = GenerateList(10);
            list.Any(e => e % 2 == 0);
        }

        [Benchmark]
        public void WhereAny()
        {
            var list = GenerateList(10);
            list.Where(e => e % 2 == 0).Any();
        }

        public List<int> GenerateList(int size)
        {
            var list = new List<int>();
            for (int count = 0; count < size; count++)
            {
                list.Add(random.Next());
            }
            return list;
        }

    }
}
```

|   Method |     Mean |   Error |  StdDev |  Gen 0 | Allocated |
|--------- |---------:|--------:|--------:|-------:|----------:|
|      Any | 108.7 ns | 1.26 ns | 1.05 ns | 0.0612 |     256 B |
| WhereAny | 125.3 ns | 2.47 ns | 3.03 ns | 0.0687 |     288 B |


The improvement really is minimal, but technically still here. 
It doesn't get optimized away.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs [hunk 5]
```diff
@@ -746,9 +746,9 @@ internal void ClearPackageLevelGrouping()
         internal void AddPackageLevelGrouping()
         {
             ItemsView.Refresh();
-            if (ItemsView.OfType<PackageItemViewModel>()
-                    .Where(p => p.PackageLevel == PackageLevel.Transitive)
-                    .Any() == true)
+            if (ItemsView
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:749]: Is the `== true` needed?

**Review by nkolev92 — COMMENTED** [review]
Looks good to me. Unfortunately most of the changes are tests, so we won't be gaining too much :)

**Comment by erdembayar** [comment]
Please rebase with latest dev branch. Current build is failing due to CI pipeline issue.

**Comment by SiliconA-Z** [comment]
Done! @erdembayar

**Comment by erdembayar** [comment]
@AtariDreams 
Thank you.
It looks like almost ready. Please create new bug issue from [here](https://github.com/NuGet/Home/issues/new/choose) and link into description of PR, please check other PRs as example.

**Review by nkolev92 — APPROVED** [review]

**Comment by SiliconA-Z** [comment]
> @AtariDreams Thank you. It looks like almost ready. Please create new bug issue from [here](https://github.com/NuGet/Home/issues/new/choose) and link into description of PR, please check other PRs as example.

Not sure how I could file a bug for this considering the behavior is not observably wrong.

**Comment by erdembayar** [comment]
> > @AtariDreams Thank you. It looks like almost ready. Please create new bug issue from [here](https://github.com/NuGet/Home/issues/new/choose) and link into description of PR, please check other PRs as example.
> 
> Not sure how I could file a bug for this considering the behavior is not observably wrong.

Just describe what you're changing, mention about potential performance improvement. It doesn't have to be long.

**Comment by SiliconA-Z** [comment]
Done: [12193](https://github.com/NuGet/Home/issues/12193)

**Review by erdembayar — APPROVED** [review]
Thank you for your contribution.

**Comment by erdembayar** [comment]
> Done: [12193](https://github.com/NuGet/Home/issues/12193)

I edited the issue little bit, because bug issue details go into 6.5 release notes in the future.
